subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -58,96 +59,109 @@
  *   - remember the current watched root project directories as old root directories for the next build
  *   - remove all non-watched root project directories from the old root directories.
  */
-public class HierarchicalFileWatcherUpdater implements FileWatcherUpdater {
+public class HierarchicalFileWatcherUpdater extends AbstractFileWatcherUpdater {
     private static final Logger LOGGER = LoggerFactory.getLogger(HierarchicalFileWatcherUpdater.class);
 
-    private final Multimap<String, Path> trackedDirectoriesForSnapshot = HashMultimap.create();
-
     private final Set<Path> watchedHierarchies = new HashSet<>();
+    private final FileWatcher fileWatcher;
 
-    private final Set<Path> knownRootProjectDirectoriesFromCurrentBuild = new HashSet<>();
-    private final Set<Path> watchedRootProjectDirectoriesFromPreviousBuild = new HashSet<>();
-
-    private final FileWatcher watcher;
     private final FileSystemLocationToWatchValidator locationToWatchValidator;
 
-    public HierarchicalFileWatcherUpdater(FileWatcher watcher, FileSystemLocationToWatchValidator locationToWatchValidator) {
-        this.watcher = watcher;
+    private final Set<Path> hierarchiesToWatchFromCurrentBuild = new HashSet<>();
+    private final Set<Path> watchedHierarchiesFromPreviousBuild = new HashSet<>();
+
+    public HierarchicalFileWatcherUpdater(FileWatcher fileWatcher, FileSystemLocationToWatchValidator locationToWatchValidator, Predicate<String> watchFilter) {
+        super(watchFilter);
+        this.fileWatcher = fileWatcher;
         this.locationToWatchValidator = locationToWatchValidator;
     }
 
     @Override
-    public void changed(Collection<CompleteFileSystemLocationSnapshot> removedSnapshots, Collection<CompleteFileSystemLocationSnapshot> addedSnapshots) {
-        removedSnapshots.forEach(snapshot -> trackedDirectoriesForSnapshot.removeAll(snapshot.getAbsolutePath()));
-        addedSnapshots.forEach(snapshot -> {
-            ImmutableList<Path> directoriesToWatch = SnapshotWatchedDirectoryFinder.getDirectoriesToWatch(snapshot);
-            trackedDirectoriesForSnapshot.putAll(snapshot.getAbsolutePath(), directoriesToWatch);
-        });
-        determineAndUpdateWatchedHierarchies();
+    public void changed(Collection<CompleteFileSystemLocationSnapshot> removedSnapshots, Collection<CompleteFileSystemLocationSnapshot> addedSnapshots, SnapshotHierarchy root) {
```

Comment:
I have some uneasy feeling about passing `root` round like this. Is that necessary?

Reply:
The hierarchical updater does not use the diff to determine the watched hierarchies, but the whole root. It only needs a trigger to re-determine what to watch when something changed in the VFS.

I don't feel uneasy to pass around the root, since the root is immutable. Do you have a better suggestion?

Reply:
I removed the root from the `SnapshotDiffListener` interface. It will still be passed to the updater.